### PR TITLE
GCMemcard: Read comments, banners, and icons via logical data offsets instead of physical ones.

### PR DESCRIPTION
**Pull request by AdmiralCurtiss**
Previously the code assumed that the data for the comments, banners, and icons was contiguous from the first block of the data's physical offset in the memory card, rather than referring to the logical position within the referenced save file, which is not only incorrect for the format but could also lead to out-of-bounds memory accesses. This should now be rectified.

Fixes:
https://bugs.dolphin-emu.org/issues/10600
https://bugs.dolphin-emu.org/issues/11820

Please scrutinize this, I'm not entirely sure if this is all correct. It seems to give correct results for all memory cards I have access to, but considering how long this old approach seemingly worked who knows.

~~The Icon/Banner code could probably use some additional rewriting on top of this to be honest, it's a mess. It also seems to disagree with itself on how the bnr_format should actually be interpreted...~~ I've done this now.

### CODE REVIEW

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 1]
```diff
@@ -536,43 +536,91 @@ u16 GCMemcard::DEntry_BlockCount(u8 index) const
...
+
+    offset_in_current_block = 0;
+    current_block = bat.GetNextBlock(current_block);
+    if (current_block < MC_FST_BLOCKS || current_block >= block_max)
+      return std::nullopt;
   }
-  return std::string((const char*)m_data_blocks[DataBlock].m_block.data() + Comment1,
-                     DENTRY_STRLEN);
+
+  return result;
```
Comment (lioncash) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: This should be std::moved, otherwise it'll get copied, given the return type isn't `std::vector<u8>`. Another way to make this more explicit is:

```cpp
return {std::move(result)};
```

or

```cpp
return std::make_optional(std::move(result));
```

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 2]
```diff
@@ -536,43 +536,91 @@ u16 GCMemcard::DEntry_BlockCount(u8 index) const
...
+  const u32 address = GetActiveDirectory().m_dir_entries[index].m_comments_address;
+  if (address == 0xFFFFFFFF)
+    return std::nullopt;
+
+  const auto data = GetSaveDataBytes(index, address, DENTRY_STRLEN * 2);
+  if (!data || data->size() != DENTRY_STRLEN * 2)
+    return std::nullopt;
+
+  auto const string_decoder = IsShiftJIS() ? SHIFTJISToUTF8 : CP1252ToUTF8;
+  auto strip_null = [](const std::string s) {
```
Comment (lioncash) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]:
```suggestion
  const auto strip_null = [](const std::string& s) {
```

File: Source/Core/Core/HW/GCMemcard/GCMemcard.h [hunk 3]
```diff
@@ -411,9 +411,12 @@ class GCMemcard
...
   u16 DEntry_BlockCount(u8 index) const;
-  u32 DEntry_CommentsAddress(u8 index) const;
-  std::string GetSaveComment1(u8 index) const;
-  std::string GetSaveComment2(u8 index) const;
+
+  std::optional<std::vector<u8>>
+  GetSaveDataBytes(u8 save_index, size_t offset = 0,
+                   size_t length = std::numeric_limits<size_t>::max()) const;
+
+  std::optional<std::pair<std::string, std::string>> GetSaveComments(u8 index) const;
```
Comment (lioncash) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.h]: This should probably have a comment indicating what each string is so others don't need to read the code to figure out its intended to be a `{title, comments}` pair.

File: Source/Core/Core/HW/GCMemcard/GCMemcard.h [hunk 4]
```diff
@@ -411,9 +411,12 @@ class GCMemcard
   u16 DEntry_FirstBlock(u8 index) const;
   // get file length in blocks
   u16 DEntry_BlockCount(u8 index) const;
-  u32 DEntry_CommentsAddress(u8 index) const;
-  std::string GetSaveComment1(u8 index) const;
-  std::string GetSaveComment2(u8 index) const;
+
+  std::optional<std::vector<u8>>
+  GetSaveDataBytes(u8 save_index, size_t offset = 0,
+                   size_t length = std::numeric_limits<size_t>::max()) const;
```
Comment (lioncash) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.h]: `<limits>` needs to be included.

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 5]
```diff
@@ -1162,120 +1162,115 @@ u32 GCMemcard::ReadAnimRGBA8(u8 index, u32* buffer, u8* delays) const
...
+  // decode frame count and formats
+  const u32 pixels_per_frame = 32 * 32;
+  const u32 palette_indices = 256;
+  u32 data_length = 0;
+  int frame_count = 0;
+  std::array<int, 8> frame_formats;
+  std::array<u8, 8> frame_delays;
+  std::array<u32, 8> frame_offsets;
+  bool has_shared_palette = false;
+  for (int i = 0; i < 8; ++i)
```
Comment (BhaaLseN) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: Something tells me the `8` here (and above in the `std::array` declarations) should be a constant...not immediately sure where it comes from though.

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 6]
```diff
@@ -1110,120 +1162,115 @@ u32 GCMemcard::ReadAnimRGBA8(u8 index, u32* buffer, u8* delays) const
...
-  {
+  u32 image_offset = GetActiveDirectory().m_dir_entries[index].m_image_offset;
+  if (image_offset == 0xFFFFFFFF)
     return 0;
-  }
-
-  u8* animData = (u8*)(m_data_blocks[DataBlock].m_block.data() + DataOffset);
 
-  switch (bnrFormat)
+  switch (bnr_format)
```
Comment (BhaaLseN) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: I wonder if we should at least document in some way that the values 0 (which is "no banner" iirc?) and 3 (which should be CI8 with unique palette?) don't affect the `image_offset`...altho, since both 1 and 3 use a palette it feels like both might need some offsetting? Is this actually correct?

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 7]
```diff
@@ -1060,34 +1109,37 @@ bool GCMemcard::ReadBannerRGBA8(u8 index, u32* buffer) const
...
-  {
+  const u32 offset = GetActiveDirectory().m_dir_entries[index].m_image_offset;
+  if (offset == 0xFFFFFFFF)
     return false;
-  }
 
-  const int pixels = 96 * 32;
+  const u32 pixel_count = 96 * 32;
+  const u32 palette_indices = 256;
+  const bool is_ci8 = (bnr_format & 1) != 0;
```
Comment (BhaaLseN) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: I wonder if this is really 4 possible values (0..3) and not just two bit flags (first one for 8/16 bit, second one for direct/indexed color; with both zero indicating a special "none" value).
YAGCD documents the former, but how we use it here practically means the values 1 and 3 aren't any different (which also ties into the comment down below for the animation).

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp (hunk 5 above)
Comment (AdmiralCurtiss) [reply Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: It's the maximum amount of frames in an icon animation, which results from the 16 bit values storing the per-frame format and delay at 2 bits each. Should probably be a constant, yeah.

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp (hunk 6 above)
Comment (AdmiralCurtiss) [reply Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: Yeah this is sketchy, I agree. Is there any banner that uses the '3' format? Maybe we should craft our own and see how the GC BIOS interprets it...

Comment (AdmiralCurtiss) [reply Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: Okay so I played around with this in the GC BIOS and as far as I can tell:

- The first two bits of the value at 0x7 (`m_banner_and_icon_flags`) are used for the banner format.
- '0' means no banner.
- '1' means paletted (8 bit per pixel + 16 bit color palette)
- '2' means direct color (16 bit per pixel)
- '3' acts the same as '0', no banner.

So this code in `ReadAnimRGBA8()` for the offset is actually correct, strange as it seems. The code in `ReadBannerRGBA8()` is wrong, as it produces wrong results in the '3' case -- except in the case of 0xFB, which happens to give the correct result of 'no banner'.

Comment (AdmiralCurtiss) [reply Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: Okay, I updated the code with my findings and left some comments. Does this look understandable to you now?

Comment (BhaaLseN) [reply Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: Seems fine! Maybe we could peek into the GC BIOS to gain more insight, but this is a good start for now.

Comment (AdmiralCurtiss) [reply Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: GC BIOS seems to agree -- in the PAL BIOS, look at 0x8131cf48 to see it mask off the last two bits and then branch depending on the four different possibilities of the value, with the branches for 0 and 3 both going to 0x8131D044.

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 8]
```diff
@@ -1178,130 +1225,122 @@ u32 GCMemcard::ReadAnimRGBA8(u8 index, u32* buffer, u8* delays) const
...
-    if (fmts[i] != 0)
+    std::memcpy(shared_palette.data(), save_data_bytes->data() + shared_palette_offset,
+                2 * palette_indices);
+  }
+
+  for (int i = 0; i < frame_count; ++i)
+  {
+    // If speed is set but there's no actual icon show the next valid icon.
+    // This is used to reduce animation speed in Pikmin and Luigi's Mansion for example.
+    // TODO: Does this really show the *next* icon instead of leaving the previous on-screen?
```
Comment (DacoTaco) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: both games have a setup of [frame][3 x blank][frame][2 x blank][frame] ( 0x4041) , ping pong style , every frame lasting 12 frames (its delay settings is 0xFFFF)
for luigi's mansion this is luigi turning his head every ~2s ( in the GC's BIOS )
it has 2 frames with his head to the left (frame 1 & 5 ) and 1 to the right ( frame 8 )

if a blank is the previous frame, in 1 ping pong it would show the left 12 frames ( * 12 frames per fps, 2.4s)  and right only 3 ( 0.6s ).
showing next frame would result in left 9 ( 1,8s ) and right 6 ( 1,2s ).
there is still something im missing as 1.2s right is nowhere near the 2s im seeing in BIOS, but with the other PR's together it seems to work out if its next frame

**Comment by AdmiralCurtiss** [comment]
Rebased this but please don't merge yet, I want to do more tests with how the 'empty frame' Pikmin/Luigi's Mansion icons actually work.

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 9]
```diff
@@ -1187,130 +1234,122 @@ u32 GCMemcard::ReadAnimRGBA8(u8 index, u32* buffer, u8* delays) const
...
 
-  int fmts[8];
-  u8* data[8];
-  int frames = 0;
-
-  for (int i = 0; i < 8; i++)
+  // decode frame count and formats
+  const int max_frame_count = 8;
+  const u32 pixels_per_frame = 32 * 32;
+  const u32 palette_indices = 256;
```
Comment (BhaaLseN) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: These could/should probably be `constexpr`, don't you think?

**Comment by AdmiralCurtiss** [comment]
[Here's a test memory card for the skipped frame behavior.](https://bugs.dolphin-emu.org/attachments/7600/skipped-frames-test-memcard.7z)

Just based on the visual output of the GC BIOS: 'Skipped' frames (format == 0) decode as the next available frame. If there is no next available frame, it will be treated as a fully transparent frame instead. One special exception is if the first frame's format is 0, the entire icon is skipped and it will just show a transparent icon instead, even if subsequent frames have a format.

**Comment by AdmiralCurtiss** [comment]
Okay, I *think* this should be all correct and readable now, I'll open this for review again.

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 10]
```diff
@@ -579,43 +579,92 @@ u16 GCMemcard::DEntry_BlockCount(u8 index) const
...
+  if (!m_valid || save_index >= DIRLEN)
+    return std::nullopt;
 
-  return GetActiveDirectory().m_dir_entries[index].m_comments_address;
-}
+  const DEntry& entry = GetActiveDirectory().m_dir_entries[save_index];
+  const BlockAlloc& bat = GetActiveBat();
+  const u16 block_count = entry.m_block_count;
+  const u16 first_block = entry.m_first_block;
+  const u32 block_max = BAT_SIZE + MC_FST_BLOCKS;
```
Comment (Warepire) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]:
```suggestion
  constexpr u32 block_max = BAT_SIZE + MC_FST_BLOCKS;
```

Comment (AdmiralCurtiss) [reply Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: Actually, looking at that again, I probably want `block_max` to be the maximum block of the current memory card, adjusting for actual card size, not the theoretical maximum.

**Comment by JMC47** [comment]
This seems to be fine for me, but I'd like confirmation from a reviewer on the code before merging.

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 11]
```diff
@@ -579,6 +579,66 @@ u16 GCMemcard::DEntry_BlockCount(u8 index) const
...
+    if (current_block < MC_FST_BLOCKS || current_block >= block_max)
+      return std::nullopt;
+  }
+
+  // then copy one block at a time into the result vector
+  while (true)
+  {
+    const GCMBlock& block = m_data_blocks[current_block - MC_FST_BLOCKS];
+    const size_t bytes_in_current_block_left =
+        std::min(BLOCK_SIZE - offset_in_current_block, static_cast<size_t>(BLOCK_SIZE));
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: Looking at this once more, I think this can be simplified to just `const size_t bytes_in_current_block_left = BLOCK_SIZE - offset_in_current_block`.

**Comment by AdmiralCurtiss** [comment]
Did a bit more cleanup. I hope it's now abundantly clear from the code how the icons and banners are stored and to be interpreted.

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 12]
```diff
@@ -579,6 +579,65 @@ u16 GCMemcard::DEntry_BlockCount(u8 index) const
...
+
+  // then copy one block at a time into the result vector
+  while (true)
+  {
+    const GCMBlock& block = m_data_blocks[current_block - MC_FST_BLOCKS];
+    const size_t bytes_in_current_block_left = BLOCK_SIZE - offset_in_current_block;
+    const size_t bytes_in_current_block_left_to_copy =
+        std::min(bytes_remaining, bytes_in_current_block_left);
+
+    for (u32 i = 0; i < bytes_in_current_block_left_to_copy; ++i)
```
Comment (leoetlino) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: Any reason you are doing a byte-by-byte insertion instead of making a result vector of the appropriate size (instead of just reserving) and using std::copy? Or using `result.insert`?

Comment (AdmiralCurtiss) [reply Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: Not particularly. I don't mind either way if you want me to change it.

Comment (leoetlino) [reply Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: Using insert should be slightly more efficient, though it most likely won't make any difference in practice. I was just wondering whether there was a reason for writing the loop manually ;)

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 13]
```diff
@@ -1228,139 +1228,143 @@ std::optional<std::vector<u32>> GCMemcard::ReadBannerRGBA8(u8 index) const
...
+        Common::DecodeCI8Image(output_frame.image_data.data(),
+                               save_data_bytes->data() + frame_offsets[j], shared_palette.data(),
+                               MEMORY_CARD_ICON_WIDTH, MEMORY_CARD_ICON_HEIGHT);
         break;
-      case CI8:  // CI8 with own palette
-        const u16* paldata = reinterpret_cast<u16*>(data[i] + 32 * 32);
-        Common::DecodeCI8Image(buffer, data[i], paldata, 32, 32);
-        buffer += 32 * 32;
+      }
+      else if (frame_formats[j] == MEMORY_CARD_ICON_FORMAT_RGB5A3)
```
Comment (leoetlino) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: `else` isn't necessary after `break`

File: Source/Core/Core/HW/GCMemcard/GCMemcard.cpp [hunk 14]
```diff
@@ -1228,139 +1228,143 @@ std::optional<std::vector<u32>> GCMemcard::ReadBannerRGBA8(u8 index) const
...
         break;
       }
-    }
-    else
-    {
-      // Speed is set but there's no actual icon
-      // This is used to reduce animation speed in Pikmin and Luigi's Mansion for example
-      // These "blank frames" show the next icon
-      for (int j = i; j < 8; ++j)
+      else if (frame_formats[j] == MEMORY_CARD_ICON_FORMAT_CI8_UNIQUE_PALETTE)
```
Comment (leoetlino) [inline Source/Core/Core/HW/GCMemcard/GCMemcard.cpp]: ditto
